Use an enum to describe directionality of RTP Stream

dontcallmedom opened the pull request:
part of #375, basd on @martinthomson's [suggestion](https://github.com/w3c/webrtc-pc/issues/375#issuecomment-163433990)

stefhak commented:
I guess it is the directionality of the RTPTransceiver (aka m-line) that is described. Regardless, this LGTM.

aboba commented:
The RTCRtpSender is not involved in receiving RTP, that is the RTCRtpReceiver. A potential alternative: 

<dl class="idl" title="enum RTCRtpDirectionality">
 <dt>sendrecv</dt>
 <dd>The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpSender</code></a>
will offer to send RTP, and will send RTP (<var>active</var>  set to "true") if the remote peer accepts.  The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpReceiver</code></a> will offer to receive RTP, and will receive RTP (<var>active</var>  set to "true") if the remote peer accepts.</dd>
 
<dt>sendonly</dt>
<dd>The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpSender</code></a>
will offer to send RTP, and will send RTP (<var>active</var>  set to "true") if the remote peer accepts.  The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpReceiver</code></a> will not offer to receive RTP, and will not receive RTP (<var>active</var> set to "false").</dd>
 
<dt>recvonly</dt>
<dd>The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpSender</code></a>
will not offer to send RTP, and will not send RTP (<var>active</var>  set to "false"). The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpReceiver</code></a>
will offer to receive RTP, and will receive RTP (<var>active</var>  set to "true") if the remote peer accepts.</dd>
 
<dt>inactive</dt>
<dd>The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpSender</code></a>
will not offer to send RTP, and will not send RTP (<var>active</var>  set to "false").
The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpReceiver</code></a> will not offer to receive RTP, and will not receive RTP (<var>active</var> set to "false").</dd>
</dl>

dontcallmedom commented:
oops, of course you're right @aboba. Would you mind updating the PR directly since I've pushed it as a branch on the shared repo? otherwise, I'll try to get to it tomorrow.

aboba commented:
Updated the PR directly.

alvestrand commented:
@pthatcherg can you see if this looks good, or would you rather stay with the booleans?
I remember some discussion here....

elagerway commented:
@pthatcherg would you have time to respond?

pthatcherg commented on webrtc.html:
```diff
@@ -3147,27 +3147,8 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
-            createOffer will be <code>sendonly</code> or <code>inactive</code>).</p>
-          </dd>
+          <dt>RTCRtpDirectionality directionality = "sendrecv"</dt>
+          <dd>The directionality of the RTP stream.</dd>
```
Whoops, look like I put comments in the wrong place before.  So I'll repeat.

I just thought of another way of doing this that I would like more:

enum RTCRtpActivity {
  active,
  inactive
}

RTCRtpActivity sender = "active",
RTCRtpActivity receiver = "active"


But if we decide to go with "direction", please call it "RTCRtpTransceiverDirection direction" instead of directionality.

pthatcherg replied:
Or better yet, add that "or boolean" convenience trick from PR 471 and use it here:
RTCRtpActivity send = "active",
RTCRtpActivity receive = "active"


Then {send: false} will still work, which would be nice.

pthatcherg commented on webrtc.html:
```diff
@@ -3179,6 +3160,32 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
+          <dd>The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpSender</code></a> will not offer to send RTP,
+            and will not send RTP (<var>active</var> set to "false"). The <a><code>RTCRtpTransceiver</code></a>'s
+            <a><code>RTCRtpReceiver</code></a> will offer to receive RTP, and will receive RTP (<var>active</var>
+            set to "true") if the remote peer accepts.</dd>
```
Until .activateSender() is called.

pthatcherg commented on webrtc.html:
```diff
@@ -3179,6 +3160,32 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
+            <a><code>RTCRtpReceiver</code></a> will not offer to receive RTP, and will not receive RTP
+            (<var>active</var> set to "false").</dd>
+        </dl>
+
```
Repeat: the "Simple Peer-to-peer Example with media before signaling" example needs to be updated.

dontcallmedom commented:
The [resolution proposed on the list](https://lists.w3.org/Archives/Public/public-webrtc/2016Apr/0052.html) is to integrate a revised version of this PR, but to close without merging #467 and #471

(cc @jan-ivar since he was active on the #471 discussion but not on this one)

alvestrand commented:
@dontcallmedom can you rebase & merge?

dontcallmedom commented:
@pthatcherg I have rebased and updated the pull request to use `direction` instead of `directionality` and to make the example match the new approach; but in your original comments you also mentioned adding "until activateSender() is called", which I wasn't sure where you wanted to add it.

Could you check the updated PR to see if anything is amiss?

pthatcherg commented on webrtc.html:
```diff
@@ -3153,27 +3153,8 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
-            createOffer will be <code>sendonly</code> or <code>inactive</code>).</p>
-          </dd>
+          <dt>RTCRtpTransceiverDirection direction = "sendrecv"</dt>
+          <dd>The direction of the RTP stream.</dd>
```
"direction of the RTP stream" is problematic.   In RTP taxonomy, "RTP stream" means something much different than the thing we are describing here. 

A better phrase might be "direction of the RtpTransceiver".

pthatcherg commented on webrtc.html:
```diff
@@ -3192,6 +3173,32 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
+            will offer to send RTP, and will send RTP (<var>active</var> set to "true") if the remote peer accepts.
+            The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpReceiver</code></a> will offer to receive RTP,
+            and will receive RTP (<var>active</var> set to "true") if the remote peer accepts.</dd>
+
```
Instead of having the parenthetical phrases (<var>active</var> set to "true"), it might be more clear with something like this:

<dd>The <a><code>RTCRtpTransceiver</code></a>'s <a><code>RTCRtpSender</code></a> will offer to send RTP, and will send RTP if the remote peer accepts, in which case <var>active</var> is set to "true".

pthatcherg commented:
I thought about this some more, and I think that if we have {direction: "sendrecv"} in the construction of an RtpTransceiver, we probably ought to have .setDirection instead of .activateSender.  That would be easier to specify and would allow the application to change direction in any way (sender/receiver active/inactive).  

But we can't just change that in this PR.  We'd have to propose the idea on the mailing list.  Perhaps in this PR, we should just go with what we have (not mention activateSender).

But I did notice some wording issues (in particular the use of "RTP stream").
